okhttp/src/main/java/io/grpc/okhttp/ExceptionHandlingFrameWriter.java

```diff
@@ -0,0 +1,224 @@
+    } catch (IOException e) {
+      transportExceptionHandler.onException(e);
+    }
+  }
+
+  @Override
+  public void flush() {
+    try {
+      frameWriter.flush();
+    } catch (IOException e) {
+      transportExceptionHandler.onException(e);
+    }
+  }
+
+  @Override
+  public void synStream(
+      boolean outFinished,
+      boolean inFinished,
+      int streamId,
+      int associatedStreamId,
+      List<Header> headerBlock) {
+    try {
+      frameWriter.synStream(outFinished, inFinished, streamId, associatedStreamId, headerBlock);
+    } catch (IOException e) {
+      transportExceptionHandler.onException(e);
+    }
+  }
+
+  @Override
+  public void synReply(boolean outFinished, int streamId,
+      List<Header> headerBlock) {
+    try {
+      frameWriter.synReply(outFinished, streamId, headerBlock);
+    } catch (IOException e) {
+      transportExceptionHandler.onException(e);
+    }
+  }
+
+  @Override
+  public void headers(int streamId, List<Header> headerBlock) {
+    try {
+      frameWriter.headers(streamId, headerBlock);
+    } catch (IOException e) {
+      transportExceptionHandler.onException(e);
+    }
+  }
+
+  @Override
+  public void rstStream(int streamId, ErrorCode errorCode) {
+    try {
+      frameWriter.rstStream(streamId, errorCode);
+    } catch (IOException e) {
+      transportExceptionHandler.onException(e);
+    }
+  }
+
+  @Override
+  public int maxDataLength() {
+    return frameWriter.maxDataLength();
+  }
+
+  @Override
+  public void data(boolean outFinished, int streamId, Buffer source, int byteCount) {
+    try {
+      frameWriter.data(outFinished, streamId, source, byteCount);
+    } catch (IOException e) {
+      transportExceptionHandler.onException(e);
+    }
+  }
+
+  @Override
+  public void settings(Settings okHttpSettings) {
+    try {
+      frameWriter.settings(okHttpSettings);
+    } catch (IOException e) {
+      transportExceptionHandler.onException(e);
+    }
+  }
+
+  @Override
+  public void ping(boolean ack, int payload1, int payload2) {
+    try {
+      frameWriter.ping(ack, payload1, payload2);
+    } catch (IOException e) {
+      transportExceptionHandler.onException(e);
+    }
+  }
+
+  @Override
+  public void goAway(int lastGoodStreamId, ErrorCode errorCode,
+      byte[] debugData) {
+    try {
+      frameWriter.goAway(lastGoodStreamId, errorCode, debugData);
+      // Flush it since after goAway, we are likely to close this writer.
+      frameWriter.flush();
+    } catch (IOException e) {
+      transportExceptionHandler.onException(e);
+    }
+  }
+
+  @Override
+  public void windowUpdate(int streamId, long windowSizeIncrement) {
+    try {
+      frameWriter.windowUpdate(streamId, windowSizeIncrement);
+    } catch (IOException e) {
+      transportExceptionHandler.onException(e);
+    }
+  }
+
+  @Override
+  public void close() {
+    try {
+      frameWriter.close();
+      if (socket != null) {
+        socket.close();
```

Comment:
This close shouldn't be here. It should be in AsyncSink instead. Being here would prevent writes from going on when shutting down.

It looks like you may have done this for testing. We may need to tweak how testing is plumbed.